src/seccomp/src/bin.rs

```diff
@@ -0,0 +1,567 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+mod compiler;
+mod syscall_table;
+
+use bincode::Error as BincodeError;
+use compiler::{Compiler, Error as FilterFormatError, Filter};
+pub use seccomp::BpfThreadMap;
+use serde_json::error::Error as JSONError;
+use std::collections::HashMap;
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::io::{BufReader, Read};
+use std::path::PathBuf;
+use std::process;
+use utils::arg_parser::{ArgParser, Argument, Arguments as ArgumentsBag};
+
+const SECCOMPILER_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const SUPPORTED_ARCHES: [&str; 2] = ["x86_64", "aarch64"];
+
+#[derive(Debug)]
+pub enum Error {
+    Bincode(BincodeError),
+    FileOpen(PathBuf, io::Error),
+    FileFormat(FilterFormatError),
+    InvalidArch,
+    JSON(JSONError),
+    MissingInputFile,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Bincode(ref err) => write!(f, "Bincode (de)serialization failed: {}", err),
+            FileFormat(ref err) => write!(f, "{}", err),
+            FileOpen(ref path, ref err) => write!(
+                f,
+                "{}",
+                format!("Failed to open file {:?}: {}", path, err).replace("\"", "")
+            ),
+            InvalidArch => write!(f, "Invalid arch"),
+            JSON(ref err) => write!(f, "Error parsing JSON: {}", err),
+            MissingInputFile => write!(f, "Missing input file"),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct Arguments {
+    input_file: String,
+    output_file: String,
+    target_arch: String,
+}
+
+fn build_arg_parser() -> ArgParser<'static> {
+    ArgParser::new()
+        .arg(
+            Argument::new("input_file")
+                .required(true)
+                .takes_value(true)
+                .help("File path of the JSON input."),
+        )
+        .arg(
+            Argument::new("output_file")
+                .required(false)
+                .takes_value(true)
+                .default_value("bpf.out")
+                .help("Optional path of the output file."),
+        )
+        .arg(
+            Argument::new("target_arch")
+                .required(true)
+                .takes_value(true)
+                .help("ARCH of the CPU that will be used to run the BPF program. One of x86_64/aarch64"),
+        )
+}
+
+fn get_argument_values(arguments: &ArgumentsBag) -> Result<Arguments> {
```

Comment:
This can be as well a `TryInto` implementation or a `new` constructor for a struct which encapsulates all the logic. WDYT?

Reply:
hmm, I think that the added complexity would not be justified. I feel that the procedural approach is more clear here